### PR TITLE
Zero out SW Fluxes when zenith angle is > 90 deg - 2 * eps

### PR DESCRIPTION
**Pull request by sriharshakandala**
## Purpose 
Temporarily zero out SW Fluxes when zenith angle is > 90 deg - 2 * eps.
Ideally, this should be handled upstream to reduce branching and provide better load balance.

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

**Comment by sriharshakandala** [comment]
@nefrathenrici : Please on including this PR in the next release.

File: src/rte/shortwave1scalar.jl [hunk 1]
```diff
@@ -9,13 +9,17 @@ function rte_sw_noscat_solve!(
...
     n_gpt, igpt = 1, 1
+    cos_zenith = bcs_sw.cos_zenith
     FT = eltype(op.angle_disc.gauss_Ds)
     solar_frac = FT(1)
     @inbounds begin
         ClimaComms.@threaded device for gcol in 1:ncol
-            set_flux_to_zero!(flux_sw, gcol)
-            compute_optical_props!(op, as, gcol)
-            rte_sw_noscat!(flux_sw, op, bcs_sw, igpt, n_gpt, solar_frac, gcol, nlev)
+            if cos_zenith[gcol] ≥ 2 * eps(FT)
```
Comment (szy21) [inline src/rte/shortwave1scalar.jl:17]: @tapios suggests we should be able to use `cos_zenith[gcol] > 0` here. Could you try and see if the test still works?

File: src/rte/shortwave1scalar.jl [hunk 2]
```diff
@@ -46,10 +50,17 @@ function rte_sw_noscat_solve_CUDA!(flux_sw::FluxSW, op::OneScalar, bcs_sw::SwBCs
...
     if gcol ≤ ncol
+        flux_up_sw = flux_sw.flux_up
+        flux_dn_sw = flux_sw.flux_dn
+        flux_net_sw = flux_sw.flux_net
         @inbounds begin
-            set_flux_to_zero!(flux_sw, gcol)
-            compute_optical_props!(op, as, gcol)
-            rte_sw_noscat!(flux_sw, op, bcs_sw, igpt, n_gpt, solar_frac, gcol, nlev)
+            μ₀ = bcs_sw.cos_zenith[gcol]
+            if μ₀ ≥ 2 * eps(FT)
```
Comment (szy21) [inline src/rte/shortwave1scalar.jl:58]:
```suggestion
            if μ₀ > 0
```

File: src/rte/shortwave1scalar.jl [hunk 3]
```diff
@@ -64,20 +75,41 @@ function rte_sw_noscat_solve!(
...
+    cos_zenith = bcs_sw.cos_zenith
     @inbounds begin
         for igpt in 1:n_gpt
             ClimaComms.@threaded device for gcol in 1:ncol
-                igpt == 1 && set_flux_to_zero!(flux_sw, gcol)
-                compute_optical_props!(op, as, gcol, igpt, lookup_sw, lookup_sw_cld)
-                solar_frac = lookup_sw.solar_src_scaled[igpt]
-                rte_sw_noscat!(flux, op, bcs_sw, igpt, n_gpt, solar_frac, gcol, nlev)
-                n_gpt > 1 && add_to_flux!(flux_sw, flux, gcol)
+                if cos_zenith[gcol] ≥ 2 * eps(FT)
```
Comment (szy21) [inline src/rte/shortwave1scalar.jl:90]:
```suggestion
                if cos_zenith[gcol] > 0
```

File: src/rte/shortwave1scalar.jl [hunk 4]
```diff
@@ -114,19 +145,48 @@ function rte_sw_noscat_solve_CUDA!(
...
+                    map!(x -> x, view(flux_up_sw, :, gcol), view(flux_up, :, gcol))
+                    map!(x -> x, view(flux_dn_sw, :, gcol), view(flux_dn, :, gcol))
+                else
+                    for ilev in 1:nlev
+                        flux_up_sw[ilev, gcol] += flux_up[ilev, gcol]
+                        flux_dn_sw[ilev, gcol] += flux_dn[ilev, gcol]
+                    end
+                end
+            end
+            if μ₀ < 2 * eps(FT)
```
Comment (szy21) [inline src/rte/shortwave1scalar.jl:175]:
```suggestion
            if μ₀ <= 0
```

File: src/rte/shortwave2stream.jl [hunk 5]
```diff
@@ -10,19 +10,24 @@ function rte_sw_2stream_solve!(
...
         flux_net_sw = flux_sw.flux_net
+
         ClimaComms.@threaded device for gcol in 1:ncol
-            set_flux_to_zero!(flux_sw, gcol)
-            compute_optical_props!(op, as, gcol)
-            # call shortwave rte solver
-            rte_sw_2stream!(op, src_sw, bcs_sw, flux_sw, solar_frac, igpt, n_gpt, ibnd, nlev, gcol)
-            for ilev in 1:nlev
-                flux_net_sw[ilev, gcol] = flux_up_sw[ilev, gcol] - flux_dn_sw[ilev, gcol]
+            if cos_zenith[gcol] ≥ 2 * eps(FT) # zero out columns with zenith > (π/2 - 2 * eps(FT))
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:22]:
```suggestion
            if cos_zenith[gcol] > 0 # zero out columns with zenith angle >= π/2
```

File: src/rte/shortwave2stream.jl [hunk 6]
```diff
@@ -65,15 +70,26 @@ function rte_sw_2stream_solve_CUDA!(
...
         @inbounds begin
-            set_flux_to_zero!(flux_sw, gcol)
             compute_optical_props!(op, as, gcol)
             # call shortwave rte solver
             rte_sw_2stream!(op, src_sw, bcs_sw, flux_sw, solar_frac, igpt, n_gpt, ibnd, nlev, gcol)
             for ilev in 1:nlev
                 flux_net_sw[ilev, gcol] = flux_up_sw[ilev, gcol] - flux_dn_sw[ilev, gcol]
             end
         end
+        if μ₀ < 2 * eps(FT) # zero out columns with zenith > (π/2 - 2 * eps(FT))
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:82]:
```suggestion
        if μ₀ <= 0 # zero out columns with zenith >= π/2
```

File: src/rte/shortwave2stream.jl [hunk 7]
```diff
@@ -100,34 +116,42 @@ function rte_sw_2stream_solve!(
...
-                # call rte shortwave solver
-                rte_sw_2stream!(op, src_sw, bcs_sw, flux, solar_frac, igpt, n_gpt, ibnd, nlev, gcol)
-                if igpt == 1
-                    map!(x -> x, view(flux_up_sw, :, gcol), view(flux_up, :, gcol))
-                    map!(x -> x, view(flux_dn_sw, :, gcol), view(flux_dn, :, gcol))
-                else
-                    for ilev in 1:nlev
-                        @inbounds flux_up_sw[ilev, gcol] += flux_up[ilev, gcol]
-                        @inbounds flux_dn_sw[ilev, gcol] += flux_dn[ilev, gcol]
+                if cos_zenith[gcol] ≥ 2 * eps(FT)
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:123]:
```suggestion
                if cos_zenith[gcol] > 0
```

File: src/rte/shortwave2stream.jl [hunk 8]
```diff
@@ -100,34 +116,42 @@ function rte_sw_2stream_solve!(
...
+                else
+                    set_flux_to_zero!(flux_sw, gcol)
                 end
             end
         end
 
         ClimaComms.@threaded device for gcol in 1:ncol
-            for ilev in 1:nlev
-                flux_net_sw[ilev, gcol] = flux_up_sw[ilev, gcol] - flux_dn_sw[ilev, gcol]
+            if cos_zenith[gcol] ≥ 2 * eps(FT)
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:151]:
```suggestion
            if cos_zenith[gcol] > 0
```

File: src/rte/shortwave2stream.jl [hunk 9]
```diff
@@ -205,6 +230,14 @@ function rte_sw_2stream_solve_CUDA!(
                     end
                 end
             end
+            if μ₀ < 2 * eps(FT) # zero out columns with zenith > (π/2 - 2 * eps(FT))
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:233]:
```suggestion
            if μ₀ <= 0 # zero out columns with zenith angle >= π/2
```

File: test/read_rfmip_clear_sky.jl [hunk 10]
```diff
@@ -33,7 +33,7 @@ function setup_rfmip_as(
     #--------------------------------------------------------------
     zenith = Array{FT, 1}(deg2rad .* Array(ds_lw_in["solar_zenith_angle"]))
     irrad = Array{FT, 1}(Array(ds_lw_in["total_solar_irradiance"]))
-    # block out coluumns with zenith > π/2
+    # block out columns with zenith > π/2
     usecol = BitArray(undef, ncol)
     usecol .= 1
     for i in 1:ncol
```
Comment (szy21) [inline test/read_rfmip_clear_sky.jl:39]: We should be able to remove this check now?

File: src/rte/RTESolver.jl [hunk 11]
```diff
@@ -113,7 +113,7 @@ solve_sw!(
     max_threads::Int,
     lookup_sw::LookUpSW,
     lookup_sw_cld::Union{LookUpCld, PadeCld, Nothing} = nothing,
-) = rte_sw_noscat_solve!(device, fluxb_sw, flux_sw, op, bcs_sw, max_threads, as, slv.lookup_sw, slv.lookup_sw_cld)
+) = rte_sw_noscat_solve!(device, fluxb_sw, flux_sw, op, bcs_sw, max_threads, as, slv.lookup_sw)
```
Comment (szy21) [inline src/rte/RTESolver.jl:116]: why don't we need the lookup_sw_cld anymore?

File: src/rte/shortwave1scalar.jl (hunk 1 above)
Comment (sriharshakandala) [reply src/rte/shortwave1scalar.jl:17]: Yes. This works. I will revise the tolerances accordingly!

File: src/rte/shortwave1scalar.jl [hunk 12]
```diff
@@ -114,19 +145,48 @@ function rte_sw_noscat_solve_CUDA!(
...
     as::AtmosphericState,
     lookup_sw::LookUpSW,
-    lookup_sw_cld::Union{LookUpCld, PadeCld, Nothing} = nothing,
 )
     gcol = threadIdx().x + (blockIdx().x - 1) * blockDim().x # global id
     nlev = nlay + 1
     n_gpt = length(lookup_sw.solar_src_scaled)
     # setting references for flux_sw
     if gcol ≤ ncol
-        set_flux_to_zero!(flux_sw, gcol)
```
Comment (szy21) [inline src/rte/shortwave1scalar.jl:124]: We don't need this anymore?

File: src/rte/shortwave2stream.jl [hunk 13]
```diff
@@ -10,19 +10,24 @@ function rte_sw_2stream_solve!(
...
         flux_net_sw = flux_sw.flux_net
+
         ClimaComms.@threaded device for gcol in 1:ncol
-            set_flux_to_zero!(flux_sw, gcol)
-            compute_optical_props!(op, as, gcol)
-            # call shortwave rte solver
-            rte_sw_2stream!(op, src_sw, bcs_sw, flux_sw, solar_frac, igpt, n_gpt, ibnd, nlev, gcol)
-            for ilev in 1:nlev
-                flux_net_sw[ilev, gcol] = flux_up_sw[ilev, gcol] - flux_dn_sw[ilev, gcol]
+            if cos_zenith[gcol] > 0 # zero out columns with zenith ≥ π/2
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:22]:
```suggestion
            if cos_zenith[gcol] > 0 # zero out columns with zenith angle ≥ π/2
```

File: src/rte/shortwave2stream.jl [hunk 14]
```diff
@@ -205,6 +230,14 @@ function rte_sw_2stream_solve_CUDA!(
                     end
                 end
             end
+            if μ₀ ≤ 0 # zero out columns with zenith ≥ π/2
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:233]:
```suggestion
            if μ₀ ≤ 0 # zero out columns with zenith angle ≥ π/2
```

File: src/rte/shortwave2stream.jl [hunk 15]
```diff
@@ -65,15 +70,26 @@ function rte_sw_2stream_solve_CUDA!(
...
         @inbounds begin
-            set_flux_to_zero!(flux_sw, gcol)
             compute_optical_props!(op, as, gcol)
             # call shortwave rte solver
             rte_sw_2stream!(op, src_sw, bcs_sw, flux_sw, solar_frac, igpt, n_gpt, ibnd, nlev, gcol)
             for ilev in 1:nlev
                 flux_net_sw[ilev, gcol] = flux_up_sw[ilev, gcol] - flux_dn_sw[ilev, gcol]
             end
         end
+        if μ₀ ≤ 0 # zero out columns with zenith ≥ π/2
```
Comment (szy21) [inline src/rte/shortwave2stream.jl:82]:
```suggestion
        if μ₀ ≤ 0 # zero out columns with zenith angle ≥ π/2
```

**Review by szy21 — APPROVED** [review]
Thanks! Could we also add a test to make sure all SW fluxes are zero when zenith angle is larger or equal than pi / 2?
